sdk/core/azure-core/inc/http/policy.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "context.hpp"
+#include "http.hpp"
+#include "transport.hpp"
+
+namespace Azure { namespace Core { namespace Http {
+
+    class NextHttpPolicy;
+
+  class HttpPolicy {
+  public:
+    // If we get a response that goes up the stack
+    // Any errors in the pipeline throws an exception
+    // At the top of the pipeline we might want to turn certain responses into exceptions
+    virtual std::unique_ptr<Response> Process(
+        Context& context,
+        Request& request,
+        NextHttpPolicy policy) const = 0;
+    virtual ~HttpPolicy() {}
+    //virtual HttpPolicy* Clone() const = 0;
+
+  protected:
+    HttpPolicy() = default;
+    HttpPolicy(const HttpPolicy& other) = default;
+    HttpPolicy(HttpPolicy&& other) = default;
+    HttpPolicy& operator=(const HttpPolicy& other) = default;
+  };
+
+  class NextHttpPolicy {
+    std::size_t m_index;
+    const std::vector<std::unique_ptr<HttpPolicy>>* m_policies;
+
+  public:
+    explicit NextHttpPolicy(const std::vector<std::unique_ptr<HttpPolicy>>* policies)
```

Comment:
In general I dislike default parameters.  Its too easy to not add them.  If we can avoid we should

Reply:
That's fine (especially since this ctor is only ever called by our own code).
So, I'd either make the index mandatory (and the piplien always passes 0).
Or, I'd have these 2 ctors:
explicit NextHttpPolicy(const std::vector<std::unique_ptr>* policies); // Assumes index=0
explicit NextHttpPolicy(const std::vector<std::unique_ptr>* policies, std::size_t index);

Reply:
Simplified to the 1 constructor